benchmarks/src/jmh/java/jetbrains/exodus/benchmark/persistit/JMHPersistItTokyoCabinetBenchmarkBase.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright 2010 - 2017 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package jetbrains.exodus.benchmark.persistit;
+
+import com.persistit.Exchange;
+import com.persistit.Persistit;
+import com.persistit.Volume;
+import com.persistit.exception.PersistitException;
+import com.persistit.logging.Slf4jAdapter;
+import java.io.IOException;
+import java.util.Properties;
+import jetbrains.exodus.ByteIterable;
+import jetbrains.exodus.ByteIterator;
+import jetbrains.exodus.benchmark.TokyoCabinetBenchmark;
+import org.jetbrains.annotations.NotNull;
+import org.junit.rules.TemporaryFolder;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.TearDown;
+import org.slf4j.LoggerFactory;
+
+abstract class JMHPersistItTokyoCabinetBenchmarkBase {
+
+    private static final ByteIterable[] successiveKeys = TokyoCabinetBenchmark.getSuccessiveEntries(TokyoCabinetBenchmark.KEYS_COUNT);
+    static final ByteIterable[] randomKeys = TokyoCabinetBenchmark.getRandomEntries(TokyoCabinetBenchmark.KEYS_COUNT);
+
+    private TemporaryFolder temporaryFolder;
+    Persistit persistit;
+    Volume volume;
+    private Exchange exchange;
+
+    public void setup() throws IOException, PersistitException {
+        TokyoCabinetBenchmark.shuffleKeys(randomKeys);
+        createEnvironment();
+    }
+
+    @TearDown(Level.Invocation)
+    public void tearDown() throws IOException, PersistitException {
+        closeDb();
+    }
+
+    void writeSuccessiveKeys(@NotNull final Exchange store) throws PersistitException {
+        for (final ByteIterable key : successiveKeys) {
+            exchange.clear();
+            ByteIterator it = key.iterator();
+            while (it.hasNext()) {
+                exchange.append(it.next());
+            }
+            exchange.getValue().put(key.getBytesUnsafe());
+            exchange.store();
+        }
+        persistit.flush();
```

Comment:
flush() is asynchronous. This is bad for the benchmark I admit, but on the other side, what do we want to measure with read/write benchmarks?
Do we want to measure "cold" read, ie reading just after the storage engine has started, and everything is read from disk? My understanding is that Persistit keep a (configurable) amount of data in memory, so that reading a value that was recently written should be faster. Maybe should I close the engine and restart it before starting the read benchmarks?
Concerning write benchmarks, it is even more complex: when can we say it is fair to consider data as written, when writing is done in the background asynchronously?